util/chunk/row_container.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunk
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/disk"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"go.uber.org/zap"
+)
+
+// RowContainer provides a place for many rows, so many that we might want to spill them into disk
+type RowContainer struct {
+	// records stores the chunks in memory.
+	records *List
+	// recordsInDisk stores the chunks in disk.
+	recordsInDisk *ListInDisk
+
+	fieldType []*types.FieldType
+	chunkSize int
+	numRow    int
+
+	// exceeded indicates that records have exceeded memQuota during
+	// this PutChunk and we should spill now.
+	// It's for concurrency usage, so access it with atomic.
+	exceeded uint32
+	// spilled indicates that records have spilled out into disk.
+	// It's for concurrency usage, so access it with atomic.
+	spilled uint32
+}
+
+// NewRowContainer creates a new RowContainer
+func NewRowContainer(fieldType []*types.FieldType, chunkSize int) RowContainer {
+	li := NewList(fieldType, chunkSize, chunkSize)
+	rc := RowContainer{records: li, fieldType: fieldType, chunkSize: chunkSize}
+	return rc
+}
+
+// NewRowContainerFromList creates a new RowContainer from chunk.List
+func NewRowContainerFromList(li *List) RowContainer {
+	rc := RowContainer{records: li, recordsInDisk: nil, fieldType: li.fieldTypes, chunkSize: li.maxChunkSize}
+	return rc
+}
+
+// NewRowContainerFromListInDisk creates a new RowContainer from chunk.ListInDisk
+func NewRowContainerFromListInDisk(li *ListInDisk) RowContainer {
```

Comment:
Why we need this function?
For unit test, you can just modify the entries in `RowContainer`.

Reply:
The reason that I did not set `c.records` to nil is because that looses the memTracker of c.records that is setup when the chunk.List is created. If we set it to nil, then next time when we call reset, it can be tricky to setup the memTracker.